### ISSUE TITLE
Change Request for Mobile Devices

### ISSUE THREAD

**Issue by thdillin** [maximejobin/rocket-nginx#41]
First of all, I am sorry if I should send this to you some other way. I looked and was unable to find a "feature request" or something similar. I will try to keep this short.

I host high volume WordPress sites. At least I consider them high volume. Most of my customers generate between 400 and 2000 new connections a minute. I have taught myself a lot about improving Nginx, PHP-fpm, and MariaDB performance over the last few months. Some of my customers use https; Varnish isn't really an option. 

I settled on the wp-rocket caching plugin because it is paid plugin supported by an active team of developers. The price is reasonable for a Dev license that allows me to use it on several sites. Although using this plugin, in the beginning, was rocky, to say the least, the team at wp-rocket has made vast improvements over the last few months. I am very happy with the plugin.

I ran across this project on git and got excited. Only to later figure out that all of my customers use the same URL for mobile and desktop clients. This forces me to rely on the PHP code inside the wp-rocket plugin to determine if the clients are mobile or desktop. Not very efficient and I think unnecessarily complex which ends up using a lot of CPU.

Which means, for all intents and purposes, your project, while impressive, doesn't provide me any benefit.

Everyone is moving toward mobile. Over 70% of my customers' traffic is generated by mobile devices. 

Here is the UGLY truth I must share. I use Plesk, and honestly I love it.  The support is good enough; it is easy to customize, and it does a lot of  work I don't have to do or write scripts to do for me. Did I mention the price is right? At any rate, Nginx is very powerful and there would be no need for any other front end caching if I could get your project to detect mobile vs. desktop user agents.

Here is my question/request.
1.  Do you think it would be possible to modify your project to include the needed logic to determine if the user agent is a mobile device or a desktop. After looking at the user agent then Nginx could grab the cached file from the appropriate location respectively.
2.  Would you mind if I modified your code and tested it to see who well it works? (I have some moderate non-critical domains I can play with)

This would provide a HUGE performance boost for my servers allowing me to scale them well beyond their current capabilities.  I use multiple nodes, but processor time is a premium cost to me. Hardware costs big cash up front, and I am not fond of "cloud". Why pay a premium for someone else to replace bad drives now and then? (I digress)

I noticed some blogs and other forums that advise people to use a mobile redirect.  This seems to add another layer of complexity.  Below the body of my message are the examples I found.  Would it be possible to start the entire rocket-Nginx configuration modification with a similar 'if' statement, where instead of a rewrite we declare and set the appropriate cache lookup directory which your project eventually tries before it sends the request over PHP?

Also, this was just an example I found as I trolled for answers, we could use the same "http_user_agent" as the wp-rocket plugin uses. In other words we could identify only mobile and default to desktop much like wp-rocket does (I think).

if ($http_user_agent ~\* "(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino") {
  set $mobile_rewrite perform;
}

if ($http_user_agent ~\* "^(1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw-(n|u)|c55\/|capi|ccwa|cdm-|cell|chtm|cldc|cmd-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc-s|devi|dica|dmob|do(c|p)o|ds(12|-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(-|_)|g1 u|g560|gene|gf-5|g-mo|go(.w|od)|gr(ad|un)|haie|hcit|hd-(m|p|t)|hei-|hi(pt|ta)|hp( i|ip)|hs-c|ht(c(-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i-(20|go|ma)|i230|iac( |-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|-[a-w])|libw|lynx|m1-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|-([1-8]|c))|phil|pire|pl(ay|uc)|pn-2|po(ck|rt|se)|prox|psio|pt-g|qa-a|qc(07|12|21|32|60|-[2-7]|i-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h-|oo|p-)|sdk\/|se(c(-|0|1)|47|mc|nd|ri)|sgh-|shar|sie(-|m)|sk-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h-|v-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl-|tdg-|tel(i|m)|tim-|t-mo|to(pl|sh)|ts(70|m-|m3|m5)|tx-9|up(.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas-|your|zeto|zte-)") {
  set $mobile_rewrite perform;
}

**Comment by maximejobin** [comment]
Don't worry about my feelings... I'm good!

I already thought about EXACTLY what you are proposing.  I also have a script to do exactly what you proposed.  But this approach did not work for me and I stopped working on it because:

1.  Checking the user agent is not enough.  You sometimes have to check other variables. And you also have to mix multiple checks to make it work.  Nginx was not made to do that and the script became unmanageable.  More than that, too many cases were left behind so it was a semi-functional feature.

2. I then found a way to simplify things using the "map" block.  The problem with that is that "map" can only be used at the http level.  That means that EVERY request will have (CSS, JS, images, ...).  That, in my opinion, would make the script useless in a sense that serving a static file that generates so much overhead is a bad idea.

I am totally open to you suggesting whatever could improve the script. My research and tests showed that creating IFs like what you provided is not efficient and is incomplete. That said, I'm open if you want to show me wrong.  I'm be glad.

That said, I honestly think that the easiest and most effective solution is to create subdomains.  It's a "one-shot" configuration that solves it all!

I could also check the script I made to do that... it was working, it was incomplete, it was creating overhead.... and it was essentially doing what you suggested. Let me check if I can find it easily.
